src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -2064,21 +2064,17 @@ Method* SystemDictionary::find_method_handle_intrinsic(vmIntrinsicID iid,
                                                        Symbol* signature,
                                                        TRAPS) {
 
-  methodHandle empty;
   const int iid_as_int = vmIntrinsics::as_int(iid);
   assert(MethodHandles::is_signature_polymorphic(iid) &&
          MethodHandles::is_signature_polymorphic_intrinsic(iid) &&
          iid != vmIntrinsics::_invokeGeneric,
          "must be a known MH intrinsic iid=%d: %s", iid_as_int, vmIntrinsics::name_at(iid));
 
-  Method** met;
+  MutexLocker ml(THREAD, InvokeMethodTable_lock);
```

Comment:
I don't think you can hold the Mutex across the entire method as the exception code below will go to Java.

Reply:
This call does not go to Java.  The other table does.

Reply:
Line 2087 can throw an exception. If you want to  hold the mutex across the entire body you need to throw the exception after releasing the lock.

Reply:
Sorry, I didn't read what you actually wrote (just what I thought you wrote). Line 2080 can also throw an exception for metaspace allocation failure.  I don't believe that we always do this outside a lock.  This is a subtle interaction that only you noticed.  I'm not sure what to do with the metaspace allocation failure.  Moving throwing the other exception outside the lock uglifies the logic but isn't hard to do.

Reply:
> This is a subtle interaction that only you noticed.

You mean in this review?  The original code clearly knew this had to be done outside the lock:

http://hg.openjdk.java.net/jdk7/jdk7/hotspot/rev/e5b0439ef4ae#l33.106

Now perhaps the current code is actually safe (as @iklam suggests) because it can't actually invoke any Java code? But that is something we would need to establish. In any case it seems an anti-pattern to allow any kind of use of CHECK from inside a locked region when we "know" we always have to throw exceptions outside of locked regions.

Reply:
The problem with the original code was the ranking and multi-purpose use of the SystemDictionary_lock.  Creating the adapters takes out a lock whose ranking is above SystemDictionary_lock.  Also the tables were shared with code that could call Java code for calling Java to create MethodType, so that code had to release the lock.  Leaking the java.lang.invoke.MethodType is ok because it'll be GC'd.

We've leaked this for many years now, maybe it's fine to keep leaking it and I should just close this and not try to fix it.

The case of throwing exceptions inside Mutex locked regions is something I thought we have all over the source code.  I couldn't find any explicit calls but the pattern of doing metaspace allocation inside of a MutexLocker is in enough places that adding an   assert(!thread->owns_locks(), "must release all locks when possibly throwing exceptions"); during in the TRAPS version of Metaspace::allocate() fails immediately.

Reply:
Right, as per our side-bar discussion, while the general rule is "no throwing exceptions while holding a VM mutex/monitor" it seems that it is safe to only throw OutOfMemoryError because `Universe::gen_out_of_memory_error` doesn't execute any Java code. It either uses one of the pre-allocated OOME instances (if available) and fills in the stacktrace, or else (or in case of secondary OOM) it uses the pre-allocated stackless singleton instance.  This should be documented somewhere of course (and perhaps even programmatically verified somehow `NoJavaCodeVerifier`? :) )

Also whilst the lower-level code makes it somewhat clear that the only exception that can arise is the metaspace OOME, it is not at all clear just by looking at a callsite like:
```
methodHandle m = Method::make_method_handle_intrinsic(iid, signature, CHECK_NULL);
```
so perhaps we even need additional macros like CHECK_OOM to verify only OOME can be thrown? (Not in this PR.)

Reply:
I added a comment above the make_method_handle_intrinsic call about why it's ok to call inside a locked region, and agree that we need better checking for the ability to run Java code from all paths with possible exceptions.  I'm not sure how to do that yet but we should file an RFE.